crates/lint/src/sol/info.rs

```diff
@@ -0,0 +1,191 @@
+use regex::Regex;
+use solar_ast::{ItemFunction, ItemStruct, VariableDefinition};
+use std::ops::ControlFlow;
+
+use crate::{
+    linter::{EarlyLintPass, LintContext},
+    sol::{
+        FunctionMixedCase, ScreamingSnakeCase, StructPascalCase, VariableMixedCase,
+        FUNCTION_MIXED_CASE, SCREAMING_SNAKE_CASE, STRUCT_PASCAL_CASE, VARIABLE_MIXED_CASE,
+    },
+};
+
+impl<'ast> EarlyLintPass<'ast> for VariableMixedCase {
+    fn check_variable_definition(
+        &mut self,
+        ctx: &LintContext<'_>,
+        var: &'ast VariableDefinition<'ast>,
+    ) -> ControlFlow<()> {
+        if var.mutability.is_none() {
+            if let Some(name) = var.name {
+                let name = name.as_str();
+                if !is_mixed_case(name) {
+                    ctx.emit(&VARIABLE_MIXED_CASE, var.span);
+                }
+            }
+        }
+        ControlFlow::Continue(())
+    }
+}
+
+impl<'ast> EarlyLintPass<'ast> for FunctionMixedCase {
+    fn check_item_function(
+        &mut self,
+        ctx: &LintContext<'_>,
+        func: &'ast ItemFunction<'ast>,
+    ) -> ControlFlow<()> {
+        if let Some(name) = func.header.name {
+            let name = name.as_str();
+            if !is_mixed_case(name) && name.len() > 1 {
+                ctx.emit(&FUNCTION_MIXED_CASE, func.body_span);
+            }
+        }
+        ControlFlow::Continue(())
+    }
+}
+
+impl<'ast> EarlyLintPass<'ast> for ScreamingSnakeCase {
+    fn check_variable_definition(
+        &mut self,
+        ctx: &LintContext<'_>,
+        var: &'ast VariableDefinition<'ast>,
+    ) -> ControlFlow<()> {
+        if let Some(mutability) = var.mutability {
+            if mutability.is_constant() || mutability.is_immutable() {
+                if let Some(name) = var.name {
+                    let name = name.as_str();
+                    if !is_screaming_snake_case(name) && name.len() > 1 {
+                        ctx.emit(&SCREAMING_SNAKE_CASE, var.span);
+                    }
+                }
+            }
+        }
+        ControlFlow::Continue(())
+    }
+}
+
+impl<'ast> EarlyLintPass<'ast> for StructPascalCase {
+    fn check_item_struct(
+        &mut self,
+        ctx: &LintContext<'_>,
+        strukt: &'ast ItemStruct<'ast>,
+    ) -> ControlFlow<()> {
+        let name = strukt.name.as_str();
+        if !is_pascal_case(name) && name.len() > 1 {
+            ctx.emit(&STRUCT_PASCAL_CASE, strukt.name.span);
+        }
+        ControlFlow::Continue(())
+    }
+}
+
+/// Check if a string is mixedCase
+///
+/// To avoid false positives like `fn increment()` or `uin256 counter`,
+/// lowercase strings are treated as mixedCase.
+pub fn is_mixed_case(s: &str) -> bool {
```

Comment:
We can use [`heck`](https://docs.rs/heck) to avoid regex; if not possible then I would recommend implementing the logic with string methods, like `s.starts_with(|c: char| matches!(...)) && s[1..]...`